src/core/lib/security/authorization/grpc_authorization_policy_provider.cc

```diff
@@ -40,6 +44,106 @@ StaticDataAuthorizationPolicyProvider::StaticDataAuthorizationPolicyProvider(
       deny_engine_(MakeRefCounted<GrpcAuthorizationEngine>(
           std::move(policies.deny_policy))) {}
 
+namespace {
+
+absl::StatusOr<std::string> ReadPolicyFromFile(absl::string_view policy_path) {
+  grpc_slice policy_slice = grpc_empty_slice();
+  grpc_error_handle error =
+      grpc_load_file(std::string(policy_path).c_str(), 0, &policy_slice);
+  if (error != GRPC_ERROR_NONE) {
+    absl::Status status =
+        absl::InvalidArgumentError(grpc_error_std_string(error));
+    GRPC_ERROR_UNREF(error);
+    return status;
+  }
+  std::string policy_contents(StringViewFromSlice(policy_slice));
+  grpc_slice_unref_internal(policy_slice);
+  return policy_contents;
+}
+
+gpr_timespec TimeoutSecondsToDeadline(int64_t seconds) {
+  return gpr_time_add(gpr_now(GPR_CLOCK_MONOTONIC),
+                      gpr_time_from_seconds(seconds, GPR_TIMESPAN));
+}
+
+}  // namespace
+
+absl::StatusOr<RefCountedPtr<grpc_authorization_policy_provider>>
+FileWatcherAuthorizationPolicyProvider::Create(
+    absl::string_view authz_policy_path, unsigned int refresh_interval_sec) {
+  GPR_ASSERT(!authz_policy_path.empty());
+  GPR_ASSERT(refresh_interval_sec > 0);
+  absl::Status status;
+  auto provider = MakeRefCounted<FileWatcherAuthorizationPolicyProvider>(
+      authz_policy_path, refresh_interval_sec, &status);
+  if (!status.ok()) return status;
+  return provider;
+}
+
+FileWatcherAuthorizationPolicyProvider::FileWatcherAuthorizationPolicyProvider(
+    absl::string_view authz_policy_path, unsigned int refresh_interval_sec,
+    absl::Status* status)
+    : authz_policy_path_(std::string(authz_policy_path)),
+      refresh_interval_sec_(refresh_interval_sec) {
+  gpr_event_init(&shutdown_event_);
+  // Initial read is done synchronously.
+  *status = ForceUpdate();
+  if (!status->ok()) {
+    return;
+  }
+  auto thread_lambda = [](void* arg) {
+    WeakRefCountedPtr<FileWatcherAuthorizationPolicyProvider> provider(
+        static_cast<FileWatcherAuthorizationPolicyProvider*>(arg));
+    GPR_ASSERT(provider != nullptr);
+    while (true) {
+      void* value = gpr_event_wait(
+          &provider->shutdown_event_,
+          TimeoutSecondsToDeadline(provider->refresh_interval_sec_));
+      if (value != nullptr) {
+        return;
+      }
+      absl::Status status = provider->ForceUpdate();
+      if (GRPC_TRACE_FLAG_ENABLED(grpc_sdk_authz_trace)) {
```

Comment:
@dfawley We are logging on every reload interval here. That way user can see the logs, and know when the implementation recovers from reload failure.

Reply:
It's common to disable INFO logs.

My preference would be:
- On reload error, log at warning level
- On reload success, don't log anything

It looks like the reload will currently be logged even if there was no change in value - at least in the Go PR - meaning this will just be continuous log spam.  If there are changes upon reloading, it could possibly be useful.

Reply:
If we only log for reload fails, then user won't know when we recover from error. I do agree that logging on every reload could be excessive, especially if we have small refresh interval.

Maybe we can include logic, where if the previous reload errored, and the current reload succeeded, we log the success status. Otherwise we only log the error status.

@markdroth @dfawley Wdyt?

Reply:
You could also log every time the contents change and show the results of the parsing.  You should still log at WARNING when there is an error parsing, but I don't think logging at WARNING makes sense otherwise.

Reply:
We don't actually have WARNING in C-core; we have only DEBUG, INFO, and ERROR.

I think we should log at ERROR when there's a failure, and we should log at INFO (a) when we have a success if the previous attempt was a failure and (b) whenever the contents change.

Reply:
@markdroth I updated the reload logic. We will be logging failures at ERROR level, and also included INFO logs whenever the policy used to make authorization decisions gets updated.
Additionally updated the logs, such that we log only once, like when we have multiple intervals where the policy contents are the same invalid policy, we will only be logging once, also updated the logic to avoid attempting the translation to rbac every time for the same invalid policy.